### ISSUE TITLE
[New Feature]: Improve Ballerina run/build/test performance using a daemon

### ISSUE THREAD

**Issue by manuranga** [ballerina-platform/ballerina-lang#39939]
### Description

Currently each completion starts in a new process. Reusing the same JVM should improve performance.

Scope
------

- Reusing the same demon for LS will be considered under a separate issue.
- At this stage a single demon process will only handle a single project.
- Current implementation will not be removed. Other CLI commands will keep using the existing implementation. If the communication to the demon fails, CLI will fallback to the current implementation.

### Related area

-> Compilation

**Comment by hasithaa** [comment]
How to handle multiple projects. Assume, I have 3 active projects, and I want to build these three ones after the other. How the demon behaves in this situation? 

In LS implementation we can handle multiple projects IIRC. @IMS94 Can you please comment, on how we handle this?

**Comment by sameerajayasoma** [comment]
Multiple package support is not implemented in the Project API level. I.e. multi-package projects. I assume the lang-server uses a map of `Projects` and load the Project instances based on the file path.

**Comment by IMS94** [comment]
Yes, we don't support multi package projects at Project API level. But as @sameerajayasoma said, LS keeps a map of projects (user can have several Ballerina projects in the current VSCode workspace) and use the file path to differentiate the project when required.

We do that in `BallerinaWorkspaceManager` class. I think we can use a single process to support multiple projects at the compile server level using a similar approach. We have to make sure that no 2 compilations run for the same project at the same time. In LS we have used locks per project to achieve that.

Since the client will communicate with the compiler server over a port, I think we will need a way to  do the following:
* Start a compilation for a given project
* Cancel the compilation of a project (If user press `Ctrl +C`)
* Check compilation status of a project (already running or not) -  (the project root path can be the identifier) or away to get notified when compilation is finished.
  * Will the request sent from client be blocked until the compilation completes?

Above things are in addition to checking the `compile server` status.

**Impact for Language Server**
Currently, LS keeps track of user's changes to the files in memory (there's a syncing protocol defined in language server protocol for that). Because of that, will LS be able to benefit of a `compiler server` (since the in-memory changes have to be reflected in the compile server to perform the compilation correctly) ?

**Comment by manuranga** [comment]
Thanks @IMS94 
I think having a single process will simplify things. I will use that approach.

> Cancel the compilation of a project 
Yes, I think we should use a duplex communication channel (similar to LS protocol), it can be TCP socket or a WebSocket instead of REST. Then CLI can send additional messages (such as Ctrl+C) while the compilation is happening.

> Will the request sent from client be blocked until the compilation completes?
Yes I think it makes sense reject or block parallel compilations.

> Impact for Language Server
I as pointed in scope section, eventual goal is to merge LS and this compiler server. At that point we should find a code level abstraction that preserve the current behavior for LS side (ie: compile from in memory) while also providing CLI behavior of compiling form disk.

**Comment by manuranga** [comment]
Caveat regarding the single process: I think lower passes (such as codeGen) is not tested in multi-project mode, since LS only needs pre-bir-gen side. So there might be race-conditions in that prat of the compiler.

**Comment by manuranga** [comment]
> 1. Users can have unsaved documents open in VSCode. 
I am not planing this level in integration yet, but when we do, we'll have to refactor code to see different open files depending on the path (CLI vs VSCode)

> 2.
Sorry I didn't fully understand the question. But AST not sharable between CLI and VSCode (if there are open files that are not saved), is that what you are referring ?

**Comment by manuranga** [comment]
@jclark suggested trying to communicate between the CLI and the compiler server using the LS protocol as well. I am trying this out.

**Comment by manuranga** [comment]
Progress
---------

Currently I have implemented a CLI that can talk to the backend via the language server protocol. I managed to send the `build` command as `workspace/executeCommand` instruction and execute it in the backend.

* Normal `bal build` takes 5450 ms.
* First run of the compiler server approach takes 3000 ms. (out of which backend is 1500ms)
Didn't expect this to be faster. I think it's only faster because I am skipping some steps.
* Subsequent runs of the compiler server approach takes 1400 ms. (out of which backend is 200ms)

Plan
----

1. I need to figure out all and implement the missing steps of compiler server backend (lookup central, create fat jar).

2. To support `run` we need to send program output incrementally. I think we can use LSP's new `$/progress` feature for this.

3. It is still too slow. I suspect cli is too slow since it still has all the classes. We may need to write a thinner CLI. (James suggested maybe we should look into a non-JVM language). But need to profile before doing anything.

3. Another option is to switch paths and focus on the VSCode.
Challenge 1 : Need to learn VSCode extension code.
Challenge 2 : Where to provide args?
Challenge 3 : We can provide `run` command off of the in memory buffers, but `build` needs to see on disk files. This is difficult to achieve under the current project API since it sees a virtual file system. We could autosave.
Advantage 1 : This may be more useful for the average user, assuming they use VSCode more often than CLI.
Advantage 2 : I don't need to worry about LS process management or writing a client.

**Comment by NipunaRanasinghe** [comment]
@manuranga I prefer having a compiler sever exposed via the CLI (instead of focusing only on the VSCode experience), because probably we can expect considerable performance improvements of the debug server with the proposed CLI approach. 

To provide a bit of context, right now the debugger server depends on the `bal run` command to launch the debuggee program.Therefore, if the user wants to rerun a debug session even without doing any code changes, the subsequent sessions will still go through all the compilation phases again and again, which could be avoided with this compiler sever approach. 

So once you have a working version of the compiler server, I'm happy to try integrating the same with the debug server and working on the possible improvement points.

**Comment by manuranga** [comment]
Ballerina extension has many UIs where a user can start a Ballerina application.
<img width="464" alt="Screenshot 2023-05-02 at 4 32 18 PM" src="https://user-images.githubusercontent.com/1686124/235652698-a3cd5af6-00fd-48ae-ba52-06cf9a09351f.png">
<img width="188" alt="Screenshot 2023-05-02 at 4 33 12 PM" src="https://user-images.githubusercontent.com/1686124/235652279-dd24b9f7-4973-4c49-9e70-e50c5b12d3be.png">
<img width="291" alt="Screenshot 2023-05-02 at 4 34 45 PM" src="https://user-images.githubusercontent.com/1686124/235652282-7627c966-d9b0-4cd8-ad76-3a00d595ff19.png">
<img width="149" alt="Screenshot 2023-05-02 at 4 34 51 PM" src="https://user-images.githubusercontent.com/1686124/235652293-1fe87cdb-a365-447d-8b37-50fb1f53f3d8.png">
<img width="172" alt="Screenshot 2023-05-02 at 4 35 54 PM" src="https://user-images.githubusercontent.com/1686124/235652296-1e818749-30d2-499b-bb8e-485327c21486.png">
<img width="218" alt="Screenshot 2023-05-02 at 4 37 08 PM" src="https://user-images.githubusercontent.com/1686124/235652300-70dc4f61-8d25-4af2-bd55-79538a95ba29.png">
<img width="327" alt="Screenshot 2023-05-02 at 4 37 38 PM" src="https://user-images.githubusercontent.com/1686124/235652304-998e8374-8987-42aa-91b1-3afbda2f0026.png">

But they all seem to come down to ether
1) Runing `bal` command form VSCode
2) Running the Debug Server (DS) which in turn runs the `bal` command

Finally the compiled program is executed as a yet another process.

Solution A)
Make the `bal` command use a background server, and not touch the VSCode extension.
Pro: No need to for me to touch the VSCode extension code.
Con: Add an additional layer, namely, CLI calling Compiler Server. Currently the added latency is too much for this approach to be an improvement over existing (2-3s). Maybe possible to improve using a thinner approach.

Solution B)
Compile inside the exiting LS process in above case 1, and compile inside DS in case 2.
Pro:  Faster
Con: I have to touch both LS and DS

@NipunaRanasinghe and others, I currently think solution B is the way to go. wdyt?

**Comment by manuranga** [comment]
@jclark pointed out that for above solution 2 to work I need to increase the lifetime of DS. AFAIK we start a DS every time use a start a new debugging session. Please confirm @NipunaRanasinghe. Does it make sense to keep a DS running and only connect to it for each session?

Solution C)
Compile inside the exiting LS process in above case 1, DS talks to LS in case 2.

Solution D)
Have a separate Build Server, make LS, DS, CLI all talk to it.

**Comment by NipunaRanasinghe** [comment]
> @jclark pointed out that for above solution 2 to work I need to increase the lifetime of DS. AFAIK we start a DS every time use a start a new debugging session. Please confirm @NipunaRanasinghe. Does it make sense to keep a DS running and only connect to it for each session?

@manuranga yes true. Right now we use the **[single session mode](https://microsoft.github.io/debug-adapter-protocol/overview)** approach for our DS, mainly to avoid complexities in multi session mode (e.g. having to handle multiple client sessions and remote VMs at once, requirement to implement the whole DS in a concurrent-safe way). 

- so for solution B, we will require a major rewrite in the DS side. 
- for solution C, we will have to open another connection with LS (per session?), which will make DS to be depended on LS (which is okay assuming that LS is guaranteed to be always running and stable) 
- regarding solution D, isn't it somewhat similar to solution A? (apologies If I'm missing something obvious) because if we have a separate build server and make CLI talk to it, in LS and DS we should be able to keep using CLI commands (which will talk to the build server similar to option A)
- on a side note, IMO it might be better to look for a solution which will abstract this compilation to a layer coupled with CLI (instead of implementing the same in each component like LS and DS), which will become more scalable incase we'll have more use cases similar to LS and DS in the future.

**Comment by hevayo** [comment]
@manuranga we did a PoC sometime back to see if we can use LS to build the executable the blocker we faced was that the Project API not designed to handle syntax node analysis tasks again after codegen. Have you notice the same or if so we might have to solve that before hand. Otherwise once you do the codegen you will not be able to reuse the same project in LS.

**Comment by manuranga** [comment]
> regarding solution D, isn't it somewhat similar to solution A? (apologies If I'm missing something obvious) Because if we have a separate build server and make CLI talk to it, in LS and DS we should be able to keep using CLI commands (which will talk to the build server similar to option A) right?

In Solution A, the CLI is not long-running. It has to be started each time which adds latency. Solution A will only be feasible after significant changes to CLI since current one is too slow. But we get to keep the CLI abstraction. In Solution D, LS and DS has to communicate with a per-running Build Server via TCP.

> @manuranga we did a PoC sometime back to see if we can use LS to build the executable the blocker we faced was that the Project API not designed to handle syntax node analysis tasks again after codegen. Have you notice the same or if so we might have to solve that before hand. Otherwise once you do the codegen you will not be able to reuse the same project in LS.

Haven't noticed yet, will keep an eye. If someone remembers what exactly broke it would be helpful.

After looking at above, I think Solution C is the lowest cost, others will take 1 month at least. Since getting something released is high priority, my current plan is to get Solution C case 1 as the first stage, tested and released first. We can reconsider the plan after that.

**Comment by manuranga** [comment]
I am trying to find a place to show a stop button for long running processors (such as services). I couldn't find a good [contribution-point](https://code.visualstudio.com/api/references/contribution-points). Only options I can think of is to add a whole [panel](https://code.visualstudio.com/api/ux-guidelines/panel), instead using built-in ᴏᴜᴛᴘᴜᴛ panel. This is bit of extra work but looks like it's needed.

@gigara Does that sound Ok?

**Comment by gigara** [comment]
@manuranga How about [Progress notification](https://code.visualstudio.com/api/ux-guidelines/notifications#progress-notification)?

**Comment by manuranga** [comment]
Thanks @gigara, that actually gave me even easier idea. I added a new command to stop a running program, `Ballerina: Stop`. Maybe I can even link it form the output window.

**Comment by NipunaRanasinghe** [comment]
> @jclark raised a point on CLI inputs and CLI specific features (such as reading a password). I just check and it doesn't work in the current debugger either, but in current `Run` it does. DAP spec gives a [way for the debugee to run in a CLI](https://microsoft.github.io/debug-adapter-protocol/specification#Reverse_Requests_RunInTerminal). We could do something slimier. Should it be opt-in or the default behavior?

@manuranga We have already implemented this `run-in-terminal` feature to support CLI inputs. (For instructions refer to the `terminal` attribute documentation in [Ballerina launch.json configurations](https://ballerina.io/learn/debug-ballerina-programs/#ballerina-launchjson-attributes)). However, currently the user has to opt-in for this (which is the common behaviour in other language debuggers as well), but we can easily make it the default behaviour if we want to.

> Another option is to emulate user inputs by sending them to Server over LSP, where Server will feed them into debugee via the pipe. Under this approach we still can't support full CLI features (such as reading passwords) and it's more work for the extension side.
If we use the above `run-in-terminal` feature, do we still need to think of this path? 

> We also discussed the possibility of creating Run as another (long running) debugger. In this case we will get the default debug UI for free.
IIUC this is equivalent to running the debugger without having any breakpoints, right? Looks okay but we may need to check on any performance impacts of always running the JVM in debug mode (https://stackoverflow.com/questions/3722841/side-effects-of-running-the-jvm-in-debug-mode)

**Comment by manuranga** [comment]
> We have already implemented this
Great. Opt-in is good.

> If we can use the above `run-in-terminal` feature, do we still need to think of this path?
This is to support input in the default case. Of course we can decide not to support inputs in default case and only support it in run-in-terminal.

> IIUC this is equivalent to running the debugger without having any breakpoints
No. The idea is to show as a debugger in the UI, but actually run java normally. This is just to get the floating toolbar appear in the UI.

**Comment by manuranga** [comment]
Hi @NipunaRanasinghe and others,

As the next iteration, my plan is to convert current implementation into a debugger.
I have experimented with `DebugAdapterInlineImplementation`, it is working. I am using the `Run Without Debugging` feature (which currently just debugs) to trigger fast-run now. Regular `Debug` still goes via existing path. I am planing to rewire `Run` code lens to also trigger `Run Without Debugging` command.

This gives me the floating debugging button toolbar. However I couldn't not figure out how to disable `Pause` button in the toolbar. I am setting the capabilities to say I don't support `supportSuspendDebuggee` but it still shows up.

**Comment by NipunaRanasinghe** [comment]
@manuranga apologies for the late response. 

> Hi @NipunaRanasinghe and others,
> 
> As the next iteration, my plan is to convert current implementation into a debugger. I have experimented with `DebugAdapterInlineImplementation`, it is working. I am using the `Run Without Debugging` feature (which currently just debugs) to trigger fast-run now. Regular `Debug` still goes via existing path. I am planing to rewire `Run` code lens to also trigger `Run Without Debugging` command.

I think this approach will be more promising as we'll have more control over the process we are running, with the support of DAP capabilities. 

> This gives me the floating debugging button toolbar. However I couldn't not figure out how to disable `Pause` button in the toolbar. I am setting the capabilities to say I don't support `supportSuspendDebuggee` but it still shows up.

The `supportSuspendDebuggee` flag actually indicates whether the debuggee should stay suspended(or resume the execution) when the debug server gets disconnected. 
It seems DAP doesn't support disabling the `pause` functionality via the capability registration atm. Probably we can open an ticket in [DAP Repository](https://github.com/microsoft/debug-adapter-protocol) to verify it with the maintainers.

**Comment by manuranga** [comment]
In today's call with @sameerajayasoma we summarized the solution in to 3 approaches
1) Make sure the tree is always in valid state after desugar. Same tree used for semantic api after backend.
2) Clone the tree before backend. Use one for backend and other for semantic api.
3) Throw away the tree after backend and start from a new tree for semantic api.

Currently I am exploring 3.

**Comment by manuranga** [comment]
I tried the approach 3. It solved some issues I had with before, but some Semantic APIs are not working.
![pkg](https://github.com/ballerina-platform/ballerina-lang/assets/1686124/ae4422a8-7c9b-4a7d-bb0d-aa5097f55535)
Found out cloned BLangPakage is missing a link to top level nodes. Looking into it now.

**Comment by manuranga** [comment]
Approach 3 seems promising, after changing to invaliding each document, instead the whole module. Simple service works under run fast now. In the process of creating unit tests for this.

Still having issues with multi module projects.

**Comment by sameerajayasoma** [comment]
Looks good. What is the percentage run improvement with FastRun?

**Comment by manuranga** [comment]
**nBallerina project run**
Cold start CLI = 2m 8s
Warm start CLI = 7s
Fast run = 3.8s

**hello world run**
Cold start CLI = 24s
Warm start CLI = 1.13s
Fast run = 200ms

**Comment by jclark** [comment]
What exactly is the difference between cold and warm start?

**Comment by manuranga** [comment]
Fist build or after `bal clean` it's cold. Subsequent builds are warm up to 24h.
In `2201.8.4` the warm CLI times are significantly improved.

**Comment by manuranga** [comment]
Multi-module projects work if I remove the [shrinkDocumnet fix](https://github.com/ballerina-platform/ballerina-lang/pull/41399), which is a temporary fix for [OOM issue](https://github.com/ballerina-platform/ballerina-lang/issues/41407) when working with healthcare libraries. It removes source as soon as each module is code generated. But that means Fast Run feature crashes the LS when working with healthcare libraries.
There is a chance this will be fixed by the [other fix](https://github.com/ballerina-platform/ballerina-lang/pull/41510) for the OOM issue, but it is not yet released.
